source/agora/consensus/data/UTXOSet.d

```diff
@@ -81,25 +126,116 @@ public class UTXOSet
 
         Params:
             tx = the transaction
+            height = Height of the block where UTXO will be activated
 
     ***************************************************************************/
 
-    public void updateUTXOCache (const ref Transaction tx) @safe
+    public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
     {
-        foreach (const ref input; tx.inputs)
+        if (tx.type == TxType.Freeze)
         {
-            auto utxo_hash = getHash(input.previous, input.index);
-            this.utxo_db.remove(utxo_hash);
+            UTXOSetValue utxo_value;
+            foreach (const ref input; tx.inputs)
+            {
+                if (!this.getUTXOSetValue(input.previous, input.index, utxo_value))
+                    return;
```

Comment:
These should `assert(0)`, not fail silently.

Reply:
I use log.error()

Reply:
Remove this part. I have confirmed that this is not necessary.